File: fusion/fuse.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""oc-fusion-in/1 on stdin -> oc-fusion/1 on stdout.

The Python half of the fusion boundary described in fusion/CONTRACT.md. It does
alignment and island rules; the TypeScript half does provider orchestration,
timing and transcript assembly. Diagnostics go to stderr, never stdout.

Exit codes: 0 success, 2 malformed input (a one-line JSON error on stderr).

No network is touched unless `config.llm` is set.

NOTE on `text`: one raw provider word can normalize to several tokens
("2026," -> ["2026"], "κ.λπ" -> ["κ", "λπ"]). When it does, several consecutive
output tokens carry the SAME `src`/`src_word` and the same raw `text`; their
`norm` fields differ. TS should treat a run of tokens sharing (src, src_word) as
one raw word.
"""
from __future__ import annotations

import json
import sys
from pathlib import Path

if __package__ in (None, ""):                              # direct `python3 fusion/fuse.py`
    sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
    __package__ = "fusion"

from . import chooser as CH                                # noqa: E402
from . import rules as R                                   # noqa: E402
from .chunking import ChunkConfig, plan_chunks             # noqa: E402
from .islands import column_class                          # noqa: E402
from .msa import align3, band_for, column_indices, consensus_pivot  # noqa: E402
from .normalize import NORMALIZER_REV, tokenize_words      # noqa: E402
from .policy import ENVELOPE, ENVELOPE_SHA16, POLICY, POLICY_SHA16  # noqa: E402

CODE_REV = "fusion/1"
IN_SCHEMA = "oc-fusion-in/1"
OUT_SCHEMA = "oc-fusion/1"
SYSTEM_IDS = ("scribe", "soniox", "ours")
ARMS = ("W", "rules", "policy")
AGREEMENT = {3: 1.0, 2: 0.67}


class InputError(ValueError):
    pass


# ------------------------------------------------------------------ validation
def _validate(payload) -> dict:
    if not isinstance(payload, dict):
        raise InputError("payload is not an object")
    if payload.get("schema") != IN_SCHEMA:
        raise InputError(f"schema must be {IN_SCHEMA!r}, got {payload.get('schema')!r}")
    systems = payload.get("systems")
    if not isinstance(systems, list) or len(systems) != 3:
        raise InputError("systems must be a list of exactly 3 entries")
    for want, s in zip(SYSTEM_IDS, systems):
        if not isinstance(s, dict):
            raise InputError("each system must be an object")
        if s.get("id") != want:
            raise InputError(f"systems must be ordered {list(SYSTEM_IDS)}; "
                             f"got {s.get('id')!r} where {want!r} was expected")
        words = s.get("words")
        if not isinstance(words, list):
            raise InputError(f"systems[{want}].words must be a list")
        for w in words:
            if not isinstance(w, dict) or not isinstance(w.get("raw"), str):
                raise InputError(f"systems[{want}].words[].raw must be a string")
            c = w.get("conf")
            if c is not None and not isinstance(c, (int, float)):
                raise InputError(f"systems[{want}].words[].conf must be a number or null")
    cfg = payload.get("config")
    if cfg is None:
        cfg = {}
    if not isinstance(cfg, dict):
        raise InputError("config must be an object")
    arm = cfg.get("arm", "rules")
    if arm not in ARMS:
        raise InputError(f"config.arm must be one of {list(ARMS)}, got {arm!r}")
    guard = cfg.get("guard", False)
    if not isinstance(guard, bool):
        raise InputError("config.guard must be a boolean")
    llm = cfg.get("llm")
    if llm is not None and not isinstance(llm, dict):
        raise InputError("config.llm must be an object or null")
    try:
        chunk_cfg = ChunkConfig.from_dict(cfg.get("chunking"))
    except (TypeError, ValueError) as e:
        raise InputError(f"config.chunking: {e}") from e
    return {"arm": arm, "guard": guard, "llm": llm, "chunking": chunk_cfg,
            "audio_sha256": payload.get("audio_sha256"), "systems": systems}


# ------------------------------------------------------------------- machinery
def _agreement(col, token):
    n = sum(1 for e in col if e == token)
    return AGREEMENT.get(n, 0.33)


class _Chunk:
    """One independently fused slice of the three streams."""

    def __init__(self, ranges, streams, pivot, cols, parts, tail, col_offset):
        self.ranges = ranges
        self.streams = streams
        self.pivot = pivot
        self.cols = cols
        self.colidx = column_indices(cols)
        self.parts = parts
        self.tail = tail
        self.col_offset = col_offset


def _analyse(tokens, chunk_cfg):
    chunks_plan, forced = plan_chunks(tokens, chunk_cfg)
    out, col_offset = [], 0
    for ranges in chunks_plan:
        sub = [tokens[k][ranges[k][0]:ranges[k][1]] for k in range(3)]
        if not any(sub):
            continue
        pivot = consensus_pivot(sub)
        cols = align3(sub[0], sub[1], sub[2], band=band_for(sub))
        parts, tail, _wsel, _wsrc = R.window_parts(cols, pivot)
        out.append(_Chunk(ranges, sub, pivot, cols, parts, tail, col_offset))
        col_offset += len(cols)
    seams = [c.col_offset for c in out[1:]]
    return out, forced, seams


def _llm_pass(chunks, llm_cfg):
    """Decide every policy-llm island. Returns {(chunk, island_index): record}."""
    need = []
    for ci, ch in enumerate(chunks):
        for k, isl in enumerate(ch.parts):
            p = POLICY.get(isl["cat"])
            if p and p["mode"] == "llm":
                need.append((ci, k, isl))
    decided = {}
    if not need:
        return decided
    batch = int((llm_cfg or {}).get("batch") or ENVELOPE.get("batch") or 15)
    bycat = {}
    for ci, k, isl in need:
        bycat.setdefault(isl["cat"], []).append((ci, k, isl))
    for cat, items in sorted(bycat.items()):
        instr = POLICY[cat]["instr"]
        for s in range(0, len(items), batch):
            group = items[s:s + batch]
            rows = [{"L": i["L"], "R": i["R"], "spans": i["spans"], "r12": i["r12"]}
                    for _, _, i in group]
            picks, err = CH.decide_batch(rows, instr, llm_cfg)
            for j, (ci, k, isl) in enumerate(group):
                if err:
                    decided[(ci, k)] = {"error": err}
                    continue
                toks, sysi, perr = CH.resolve(picks.get(j), isl)
                if perr:
                    decided[(ci, k)] = {"error": perr}
                else:
                    decided[(ci, k)] = {"tokens": toks, "src": sysi,
                                        "confidence": (picks.get(j) or ("", ""))[1]}
    return decided


def _island_output(isl, arm, guard, decided_rec):
    """(triples, stage, rule, guard_fired, llm_record) for one island."""
    llm_rec = None
    if arm == "W":
        tr, stage, rule = isl["vote_tr"], "rule", R.RULE_VOTE
    elif arm == "rules":
        tr, stage, rule = isl["r12_tr"], "rule", R.RULE_R12
    else:
        p = POLICY.get(isl["cat"])
        if p is None:
            tr, stage, rule = isl["vote_tr"], "rule", R.RULE_VOTE
        elif p["mode"] == "rule":
            tr, stage, rule = R.RULES[p["rule"]](isl), "rule", p["rule"]
        else:
            rec = decided_rec or {"error": "llm disabled"}
            if "error" in rec:
                llm_rec = {"error": rec["error"]}
                tr, stage, rule = isl["r12_tr"], "rule", R.RULE_R12
            else:
                llm_rec = {"confidence": rec.get("confidence")}
                stage, rule = "llm", None
                sysi, toks = rec["src"], rec["tokens"]
                if sysi is not None:
                    tr = isl["span_tr"][sysi]
                elif not toks:
                    tr = []
                else:
                    tr = isl["r12_tr"]
    fired = False
    if guard:
        # The guard replaces the island output with a whole system's span. It
        # does not change which rule routed the island, so `rule` is left alone
        # and `guard_fired` is the flag that says the text was overridden.
        tr, fired = R.apply_guard_tr(isl, list(tr))
    return list(tr), stage, rule, fired, llm_rec


# ------------------------------------------------------------------------ main
def fuse(payload: dict) -> dict:
    cfg = _validate(payload)
    arm, guard, llm_cfg = cfg["arm"], cfg["guard"], cfg["llm"]

    raws = [[w["raw"] for w in s["words"]] for s in cfg["systems"]]
    confs = [[w.get("conf") for w in s["words"]] for s in cfg["systems"]]
    tokenized = [tokenize_words(r) for r in raws]
    tokens = [t for t, _o in tokenized]
    owner = [o for _t, o in tokenized]

    chunks, forced, seams = _analyse(tokens, cfg["chunking"])
    decided = {}
    if arm == "policy" and llm_cfg:
        decided = _llm_pass(chunks, llm_cfg)

    def word_of(ch, sysi, col):
        """(raw word index, raw text, conf) behind a column entry."""
        local = ch.colidx[col][sysi]
        if local is None:
            return None, None, None
        gidx = ch.ranges[sysi][0] + local
        widx = owner[sysi][gidx]
        return widx, raws[sysi][widx], confs[sysi][widx]

    def alternatives(ch, isl):
        out = []
        for k in range(3):
            words, conf_vals = [], []
            seen = set()
            for _t, _s, col in isl["span_tr"][k]:
                widx, raw, cf = word_of(ch, k, col)
                if widx is None or widx in seen:
                    continue
                seen.add(widx)
                words.append(raw)
                if cf is not None:
                    conf_vals.append(cf)
            out.append({
                "src": SYSTEM_IDS[k],
                "text": " ".join(words),
                "norm": list(isl["spans"][k]),
                "conf": (sum(conf_vals) / len(conf_vals)) if conf_vals else None,
            })
        return out

    out_tokens, out_islands, dropped = [], [], []
    stats_cat, stats_stage = {}, {"agree": 0, "rule": 0, "llm": 0}
    n_guard = 0
    isl_counter = 0

    def emit(ch, triples, stage, island_id, alts):
        for tok, sysi, col in triples:
            widx, raw, _cf = word_of(ch, sysi, col)
            out_tokens.append({
                "i": len(out_tokens),
                "text": raw if raw is not None else tok,
                "norm": tok,
                "src": SYSTEM_IDS[sysi],
                "src_word": widx,
                "col": ch.col_offset + col,
                "island": island_id,
                "stage": stage,
                "agreement": _agreement(ch.cols[col], tok),
                "alternatives": alts,
            })
            stats_stage[stage] = stats_stage.get(stage, 0) + 1

    for ci, ch in enumerate(chunks):
        for k, isl in enumerate(ch.parts):
            emit(ch, isl["before_tr"], "agree", None, None)
            iid = f"isl_{isl_counter}"
            isl_counter += 1
            tr, stage, rule, fired, llm_rec = _island_output(
                isl, arm, guard, decided.get((ci, k)))
            n_guard += int(fired)
            alts = alternatives(ch, isl)
            emit(ch, tr, stage, iid, alts)

            srcs = {s for _t, s, _c in tr}
            chosen_src = SYSTEM_IDS[next(iter(srcs))] if len(srcs) == 1 else None
            out_islands.append({
                "id": iid,
                "cols": [ch.col_offset + isl["s"], ch.col_offset + isl["e"] - 1],
                "category": isl["cat"],
                "candidates": {"scribe": list(isl["spans"][0]),
                               "soniox": list(isl["spans"][1]),
                               "ours": list(isl["spans"][2]),
                               "vote": list(isl["vote"]),
                               "r12": list(isl["r12"])},
                "stage": stage,
                "rule": rule,
                "chosen_src": chosen_src,
                "guard_fired": fired,
                "llm": llm_rec,
            })
            stats_cat[isl["cat"]] = stats_cat.get(isl["cat"], 0) + 1

            if not tr:
                best = max(range(3), key=lambda k2: (len(isl["spans"][k2]), -k2))
                if isl["spans"][best]:
                    dropped.append({"island": iid, "src": SYSTEM_IDS[best],
                                    "text": alts[best]["text"],
                                    "rule": rule or stage})
        emit(ch, ch.tail, "agree", None, None)

    chunk_out = cfg["chunking"].as_dict()
    chunk_out.update({"n_chunks": len(chunks), "forced_cuts": forced, "seams": seams})

    return {
        "schema": OUT_SCHEMA,
        "audio_sha256": cfg["audio_sha256"],
        "config": {
            "arm": arm,
            "guard": guard,
            "policy_sha": POLICY_SHA16,
            "llm_envelope_sha": ENVELOPE_SHA16 if (arm == "policy" and llm_cfg) else None,
            "code_rev": CODE_REV,
            "normalizer_rev": NORMALIZER_REV,
            "chunking": chunk_out,
            "components": {SYSTEM_IDS[k]: cfg["systems"][k].get("params_sha")
                           for k in range(3)},
        },
        "tokens": out_tokens,
        "islands": out_islands,
        "dropped": dropped,
        "stats": {
            "n_tokens": len(out_tokens),
            "n_islands": len(out_islands),
            "by_category": stats_cat,
            "by_stage": stats_stage,
            "guard_fired": n_guard,
        },
    }


def main(argv=None) -> int:
    raw = sys.stdin.read()
    try:
        payload = json.loads(raw)
    except Exception as e:
        print(json.dumps({"error": "invalid JSON on stdin", "detail": str(e)[:200]},
                         ensure_ascii=False), file=sys.stderr)
        return 2
    try:
        out = fuse(payload)
    except InputError as e:
        print(json.dumps({"error": "malformed oc-fusion-in/1", "detail": str(e)[:400]},
                         ensure_ascii=False), file=sys.stderr)
        return 2
    json.dump(out, sys.stdout, ensure_ascii=False)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/lib/fusion/engine/fuse.ts
/**
 * `oc-fusion-in/1` in, `oc-fusion/1` out. Ported from `fusion/fuse.py`.
 *
 * Alignment and island rules live here; provider orchestration, timing and
 * transcript assembly stay where they already are, in the fusion library above
 * this directory.
 *
 * The LLM chooser is deliberately NOT ported. It contributes to no measured
 * result, it is off by default, and an external model's answers cannot be part
 * of a deterministic equivalence proof. A request that asks for it is rejected
 * here with `LlmNotSupportedError` so the caller can route that one path to the
 * Python, rather than being quietly served a different arm.
 *
 * On `text`: one raw provider word can normalize to several tokens, so several
 * consecutive output tokens can carry the same `src`/`src_word` and the same
 * raw `text` while their `norm` differs. A run of tokens sharing
 * `(src, src_word)` is one raw word.
 */
import { align3, bandFor, columnIndices, consensusPivot, type Column } from "./msa.js";
import { chunkConfigFrom, planChunks, type ChunkConfig, type ChunkRanges } from "./chunking.js";
import { columnClass } from "./islands.js";
import { NORMALIZER_REV, tokenizeWords } from "./normalize.js";
import * as R from "./rules.js";
import { type Island, type Triple } from "./rules.js";
import { POLICY_SHA16, type LoadedPolicy } from "./policy.js";
import { pySum } from "./pyjson.js";

export const CODE_REV = "fusion/1";
export const IN_SCHEMA = "oc-fusion-in/1";
export const OUT_SCHEMA = "oc-fusion/1";
export const SYSTEM_IDS = ["scribe", "soniox", "ours"] as const;
export const ARMS = ["W", "rules", "policy"] as const;

export type Arm = (typeof ARMS)[number];

export class InputError extends Error {
    constructor(message: string) {
        super(message);
        this.name = "InputError";
    }
}

export class LlmNotSupportedError extends Error {
    constructor() {
        super("the LLM chooser is not implemented in the TypeScript engine; "
            + "route this request to the Python engine");
        this.name = "LlmNotSupportedError";
    }
}

interface InputWord {
    raw: string;
    start?: number;
    end?: number;
    conf?: number | null;
}

interface InputSystem {
    id: string;
    params_sha?: string;
    words: InputWord[];
}

interface ValidatedConfig {
    arm: Arm;
    guard: boolean;
    llm: Record<string, unknown> | null;
    chunking: ChunkConfig;
    audio_sha256: unknown;
    systems: InputSystem[];
}

/** Python renders a list of strings with single quotes and ", " between them. */
function pyList(xs: readonly string[]): string {
    return "[" + xs.map((x) => `'${x}'`).join(", ") + "]";
}

/** Python renders a string with single quotes in these messages; so does this. */
function pyRepr(v: unknown): string {
    if (typeof v === "string") return `'${v.replace(/'/g, "\\'")}'`;
    if (v === null || v === undefined) return "None";
    if (typeof v === "boolean") return v ? "True" : "False";
    return String(v);
}

function validate(payload: unknown): ValidatedConfig {
    if (typeof payload !== "object" || payload === null || Array.isArray(payload)) {
        throw new InputError("payload is not an object");
    }
    const p = payload as Record<string, unknown>;
    if (p.schema !== IN_SCHEMA) {
        throw new InputError(`schema must be ${pyRepr(IN_SCHEMA)}, got ${pyRepr(p.schema)}`);
    }
    const systems = p.systems;
    if (!Array.isArray(systems) || systems.length !== 3) {
        throw new InputError("systems must be a list of exactly 3 entries");
    }
    for (let n = 0; n < 3; n++) {
        const want = SYSTEM_IDS[n];
        const s = systems[n];
        if (typeof s !== "object" || s === null || Array.isArray(s)) {
            throw new InputError("each system must be an object");
        }
        const sys = s as Record<string, unknown>;
        if (sys.id !== want) {
            throw new InputError(
                `systems must be ordered ${pyList([...SYSTEM_IDS])}; `
                + `got ${pyRepr(sys.id)} where ${pyRepr(want)} was expected`);
        }
        const words = sys.words;
        if (!Array.isArray(words)) {
            throw new InputError(`systems[${want}].words must be a list`);
        }
        for (const w of words) {
            if (typeof w !== "object" || w === null || Array.isArray(w)
                || typeof (w as Record<string, unknown>).raw !== "string") {
                throw new InputError(`systems[${want}].words[].raw must be a string`);
            }
            const c = (w as Record<string, unknown>).conf;
            if (c !== undefined && c !== null && typeof c !== "number") {
                throw new InputError(`systems[${want}].words[].conf must be a number or null`);
            }
        }
    }

    let cfg = p.config;
    if (cfg === undefined || cfg === null) cfg = {};
    if (typeof cfg !== "object" || Array.isArray(cfg)) {
        throw new InputError("config must be an object");
    }
    const c = cfg as Record<string, unknown>;

    const arm = (c.arm ?? "rules") as Arm;
    if (!ARMS.includes(arm)) {
        throw new InputError(
            `config.arm must be one of ${pyList([...ARMS])}, got ${pyRepr(c.arm)}`);
    }
    const guard = c.guard ?? false;
    if (typeof guard !== "boolean") throw new InputError("config.guard must be a boolean");

    const llm = c.llm;
    if (llm !== undefined && llm !== null
        && (typeof llm !== "object" || Array.isArray(llm))) {
        throw new InputError("config.llm must be an object or null");
    }

    let chunking: ChunkConfig;
    try {
        chunking = chunkConfigFrom(c.chunking as Record<string, unknown> | null | undefined);
    } catch (e) {
        throw new InputError(`config.chunking: ${(e as Error).message}`);
    }

    return {
        arm,
        guard,
        llm: (llm ?? null) as Record<string, unknown> | null,
        chunking,
        audio_sha256: p.audio_sha256 ?? null,
        systems: systems as InputSystem[],
    };
}

const AGREEMENT = new Map<number, number>([[3, 1.0], [2, 0.67]]);

function agreement(col: Column, token: string): number {
    let n = 0;
    for (const e of col) if (e === token) n += 1;
    return AGREEMENT.get(n) ?? 0.33;
}

interface Chunk {
    ranges: ChunkRanges;
    pivot: number;
    cols: Column[];
    colidx: [number | null, number | null, number | null][];
    parts: Island[];
    tail: Triple[];
    colOffset: number;
}

function analyse(tokens: string[][], chunkCfg: ChunkConfig): {
    chunks: Chunk[];
    forced: number;
    seams: number[];
} {
    const { chunks: plan, forced } = planChunks(tokens, chunkCfg);
    const out: Chunk[] = [];
    let colOffset = 0;
    for (const ranges of plan) {
        const sub = [0, 1, 2].map((k) => tokens[k].slice(ranges[k][0], ranges[k][1]));
        if (!sub.some((s) => s.length > 0)) continue;
        const pivot = consensusPivot(sub);
        const cols = align3(sub[0], sub[1], sub[2], bandFor(sub));
        const { islands: parts, tail_tr: tail } = R.windowParts(cols, pivot);
        out.push({ ranges, pivot, cols, colidx: columnIndices(cols), parts, tail, colOffset });
        colOffset += cols.length;
    }
    return { chunks: out, forced, seams: out.slice(1).map((c) => c.colOffset) };
}

/** Python's `max(range(3), key=...)` keeps the first index on a tie. */
function longestSpan(spans: readonly (readonly string[])[]): number {
    let best = 0;
    for (let k = 1; k < 3; k++) {
        const a: [number, number] = [spans[k].length, -k];
        const b: [number, number] = [spans[best].length, -best];
        if (a[0] > b[0] || (a[0] === b[0] && a[1] > b[1])) best = k;
    }
    return best;
}

function islandOutput(isl: Island, arm: Arm, guard: boolean, policy: LoadedPolicy): {
    tr: Triple[];
    stage: string;
    rule: string | null;
    fired: boolean;
    llm: null;
} {
    let tr: Triple[];
    let stage: string;
    let rule: string | null;

    if (arm === "W") {
        tr = isl.vote_tr; stage = "rule"; rule = R.RULE_VOTE;
    } else if (arm === "rules") {
        tr = isl.r12_tr; stage = "rule"; rule = R.RULE_R12;
    } else {
        const p = policy.policy[isl.cat];
        if (p === undefined) {
            tr = isl.vote_tr; stage = "rule"; rule = R.RULE_VOTE;
        } else if (p.mode === "rule") {
            tr = R.RULES[p.rule as string](isl); stage = "rule"; rule = p.rule as string;
        } else {
            // With no chooser, the Python falls back to R1+R2 and records the
            // reason. Reaching here means the caller asked for the LLM arm
            // without an LLM, which `fuse` has already rejected.
            throw new LlmNotSupportedError();
        }
    }

    let fired = false;
    if (guard) {
        // The guard replaces the island text with a whole system's span. It
        // does not change which rule routed the island, so `rule` stands and
        // `guard_fired` is what says the text was overridden.
        const res = R.applyGuardTr(isl, tr);
        tr = res.out;
        fired = res.fired;
    }
    return { tr: [...tr], stage, rule, fired, llm: null };
}

export interface FuseOutput {
    schema: string;
    audio_sha256: unknown;
    config: Record<string, unknown>;
    tokens: Record<string, unknown>[];
    islands: Record<string, unknown>[];
    dropped: Record<string, unknown>[];
    stats: Record<string, unknown>;
}

export function fuse(payload: unknown, policy: LoadedPolicy): FuseOutput {
    const cfg = validate(payload);
    const { arm, guard } = cfg;

    if (arm === "policy" && cfg.llm) throw new LlmNotSupportedError();

    const raws = cfg.systems.map((s) => s.words.map((w) => w.raw));
    const confs = cfg.systems.map((s) => s.words.map((w) => w.conf ?? null));
    const tokenized = raws.map((r) => tokenizeWords(r));
    const tokens = tokenized.map((t) => t.tokens);
    const owner = tokenized.map((t) => t.owner);

    const { chunks, forced, seams } = analyse(tokens, cfg.chunking);

    const wordOf = (ch: Chunk, sysi: number, col: number): {
        widx: number | null; raw: string | null; conf: number | null;
    } => {
        const local = ch.colidx[col][sysi];
        if (local === null) return { widx: null, raw: null, conf: null };
        const gidx = ch.ranges[sysi][0] + local;
        const widx = owner[sysi][gidx];
        return { widx, raw: raws[sysi][widx], conf: confs[sysi][widx] };
    };

    const alternatives = (ch: Chunk, isl: Island) => {
        const out: Record<string, unknown>[] = [];
        for (let k = 0; k < 3; k++) {
            const words: string[] = [];
            const confVals: number[] = [];
            const seen = new Set<number>();
            for (const [, , col] of isl.span_tr[k]) {
                const { widx, raw, conf } = wordOf(ch, k, col);
                if (widx === null || seen.has(widx)) continue;
                seen.add(widx);
                words.push(raw as string);
                if (conf !== null) confVals.push(conf);
            }
            out.push({
                src: SYSTEM_IDS[k],
                text: words.join(" "),
                norm: [...isl.spans[k]],
                conf: confVals.length ? pySum(confVals) / confVals.length : null,
            });
        }
        return out;
    };

    const outTokens: Record<string, unknown>[] = [];
    const outIslands: Record<string, unknown>[] = [];
    const dropped: Record<string, unknown>[] = [];
    const statsCat: Record<string, number> = {};
    const statsStage: Record<string, number> = { agree: 0, rule: 0, llm: 0 };
    let nGuard = 0;
    let islCounter = 0;

    const emit = (
        ch: Chunk, triples: readonly Triple[], stage: string,
        islandId: string | null, alts: Record<string, unknown>[] | null,
    ) => {
        for (const [tok, sysi, col] of triples) {
            const { widx, raw } = wordOf(ch, sysi, col);
            outTokens.push({
                i: outTokens.length,
                text: raw !== null ? raw : tok,
                norm: tok,
                src: SYSTEM_IDS[sysi],
                src_word: widx,
                col: ch.colOffset + col,
                island: islandId,
                stage,
                agreement: agreement(ch.cols[col], tok),
                alternatives: alts,
            });
            statsStage[stage] = (statsStage[stage] ?? 0) + 1;
        }
    };

    for (const ch of chunks) {
        for (const isl of ch.parts) {
            emit(ch, isl.before_tr, "agree", null, null);
            const iid = `isl_${islCounter}`;
            islCounter += 1;
            const { tr, stage, rule, fired } = islandOutput(isl, arm, guard, policy);
            nGuard += fired ? 1 : 0;
            const alts = alternatives(ch, isl);
            emit(ch, tr, stage, iid, alts);

            const srcs = new Set(tr.map(([, s]) => s));
            const chosenSrc = srcs.size === 1
                ? SYSTEM_IDS[srcs.values().next().value as number]
                : null;

            outIslands.push({
                id: iid,
                cols: [ch.colOffset + isl.s, ch.colOffset + isl.e - 1],
                category: isl.cat,
                candidates: {
                    scribe: [...isl.spans[0]],
                    soniox: [...isl.spans[1]],
                    ours: [...isl.spans[2]],
                    vote: [...isl.vote],
                    r12: [...isl.r12],
                },
                stage,
                rule,
                chosen_src: chosenSrc,
                guard_fired: fired,
                llm: null,
            });
            statsCat[isl.cat] = (statsCat[isl.cat] ?? 0) + 1;

            if (!tr.length) {
                const best = longestSpan(isl.spans);
                if (isl.spans[best].length) {
                    dropped.push({
                        island: iid,
                        src: SYSTEM_IDS[best],
                        text: alts[best].text,
                        rule: rule ?? stage,
                    });
                }
            }
        }
        emit(ch, ch.tail, "agree", null, null);
    }

    const chunkOut: Record<string, unknown> = {
        rev: cfg.chunking.rev,
        max_tokens: cfg.chunking.max_tokens,
        anchor_n: cfg.chunking.anchor_n,
        search_radius: cfg.chunking.search_radius,
        min_pause_fallback: cfg.chunking.min_pause_fallback,
        n_chunks: chunks.length,
        forced_cuts: forced,
        seams,
    };

    return {
        schema: OUT_SCHEMA,
        audio_sha256: cfg.audio_sha256,
        config: {
            arm,
            guard,
            policy_sha: POLICY_SHA16,
            // Python reports the envelope hash only when the LLM arm actually
            // ran. That arm is rejected above, so it is always absent here.
            llm_envelope_sha: null,
            code_rev: CODE_REV,
            normalizer_rev: NORMALIZER_REV,
            chunking: chunkOut,
            components: Object.fromEntries(
                [0, 1, 2].map((k) => [SYSTEM_IDS[k], cfg.systems[k].params_sha ?? null]),
            ),
        },
        tokens: outTokens,
        islands: outIslands,
        dropped,
        stats: {
            n_tokens: outTokens.length,
            n_islands: outIslands.length,
            by_category: statsCat,
            by_stage: statsStage,
            guard_fired: nGuard,
        },
    };
}

/** Exported for the differential harness, which needs the column view too. */
export { columnClass };
